server/rpc.go

```diff
@@ -50,6 +51,80 @@ type Broadcaster interface {
 	Sign([]byte) ([]byte, error)
 }
 
+type BroadcastSessionsManager struct {
+	broadcastSessions []*BroadcastSession
+	sessLock          *sync.Mutex
+}
+
+// func (bsm *BroadcastSessionsManager) selectFromList(node *core.LivepeerNode, pl core.PlaylistManager) *BroadcastSession {
+func (bsm *BroadcastSessionsManager) selectFromList() *BroadcastSession {
+	// bsm.checkStatusOfList(numOrchs, node, pl)
+	bsm.checkStatusOfList()
+
+	bsm.sessLock.Lock()
+	defer bsm.sessLock.Unlock()
+
+	sess, sessions := bsm.broadcastSessions[0], bsm.broadcastSessions[1:]
+	bsm.broadcastSessions = sessions
+
+	return sess
+}
+
+func (bsm *BroadcastSessionsManager) removeFromList(session *BroadcastSession) {
+	bsm.sessLock.Lock()
+	defer bsm.sessLock.Unlock()
+	var sessions []*BroadcastSession
+	for i, s := range bsm.broadcastSessions {
+		if s == session {
+			sessions = append(bsm.broadcastSessions[:i], bsm.broadcastSessions[i+1:]...)
+			break
+		}
+	}
+	bsm.broadcastSessions = sessions
+}
+
+func (bsm *BroadcastSessionsManager) addToList(sess *BroadcastSession) {
```

Comment:
Here, we would be cycling continuously through the list like a queue, selecting from the head and pushing (appending) to the tail. This is probably OK if we want to give everyone a chance at processing a segment, but for now we probably want to give more weight to quick respondents.

To reward quick responders with more segments, we want stack-like behavior: select from the head, and push back to the head (prepend rather than append).

The asymptotics of prepend in golang don't appear great (involves copying a slice which is not ideal under a lock), but I wonder if treating the end of the list as the head to would help here. So:

```golang
selectFromList() {
  last = len(broadcastSessions) - 1
  // ...do bounds checking...
  // select from end of list
  sess, sessions = broadacastSessions[last], broadcastSessions[:last]
}
addToList(session) {
  // append to end of list
  sessions = append(broadcastSessions, session)
}
```

Slice indexing behavior might totally negate this, but it looks a little better than an explicit copy.